### PR TITLE
Improve error messages in network_client

### PR DESCRIPTION
**Pull request by dantswain**
- More consistently refer to brokers
- Specify async case in error message
- Specify send/receive in sync case

@bjhaid Quick review/merge would be much appreciated as this is related to a problem we're experiencing in production.  We're getting timeouts from the network client but can't tell if it's in the send or receive step because the error messages are identical and both return `nil`.

### CODE REVIEW

**Comment by bjhaid** [comment]
Do you need an hex release for this?

**Comment by dantswain** [comment]
No I can just point at the SHA.  Thanks! :+1:

**Comment by dantswain** [comment]
@bjhaid Just wanted to say thanks again.  I was able to get this into production yesterday and it helped me debug issues overnight last night.

**Comment by bjhaid** [comment]
Did you find the problem? If yes what was it?

**Comment by dantswain** [comment]
We're still tracking that down, but I did learn some interesting things.  Sorry if not all of this brain dump is interesting to you, but it's helping me organize my own thoughts, too ;)

tl;dr version: 

I think we're experiencing some problems on our brokers that are causing timeouts.  A couple potential takeaways as they related to KafkaEx: 1) I'm finding it better in practice to use one KafkaEx worker per partition rather than sharing one across multiple partitions, 2) it may be worthwhile to expose KafkaEx.Supervisor's restart strategy parameters to the application config.

Much longer version:

* We're experiencing a timeout while reading from two of our brokers at certain times of day.
* It may be coincidence, or just another symptom of the underlying problem, but during those time frames we see a lot of ISR resizing happening on the brokers.  The ISR rebalancing is happening on different topics than the ones we are reading from with KafkaEx.
* This PR helped me isolate the timeout to the read portion of the transaction.  That's really valuable information!
* The timeouts were causing the KafkaEx worker to crash, which is fine.  The problem was that this caused a ripple effect and brought down enough linked processes that the whole VM was crashing.
* In my first iteration, I was sharing one KafkEx worker with six partition consumers.  In a gut instinct move, I changed that to one worker per partition.
  * At that point, I wasn't sure if the timeout was due to the broker being slow to respond or due to the consumer box experiencing resource contention.  I could be wrong, but my thinking was that sharing a KafkaEx worker process and a TCP/IP connection could be a source of contention, so using one worker per partition may relieve that.
  * Given what we observed last night (this issue is ongoing on our end), I'm mostly convinced that the timeouts are not due to resource contention on the consumer end.
  * The KafkaEx worker process crashing caused all of my consumer processes to crash because they were each linked to the worker process.  That number of process crashes exceeds Elixir's default restart strategy limits (3 restarts in 5 seconds iirc).  Using a separate KafkaEx worker process for each partition consumer process alleviates that because each KafkaEx worker process is only linked to a single consumer process.  I also upped the restart params in my supervisor (exposed it to config, actually).
  * It might be a good idea to expose the KafkaEx.Supervisor restart strategy parameters to the application config.  The number of allowed restarts may need to be tweaked as a function of the number of worker processes running on the user's VM.  I.e.,

```
# in config.exs
config :kafka_ex,
  max_restarts: 3,
  max_seconds: 5

# in the supervisor
@default_max_restarts 10
@default_max_seconds 60

def init(_args) do
  # ...
  supervisor_opts = [strategy: :simple_one_for_one, max_restarts: Application.get_env(:kafka_ex, :max_restarts, @default_max_restarts), max_seconds: Application.get_env(:kafka_ex, :max_seconds, @default_max_seconds)]
  # ...
end
```

This part is just my opinion; provided for the sake of discussion only - I'm not trying to say KafkaEx needs to do this.  Typically when I write library code in Elixir, I omit the supervisor and let the user (which is usually me, so take this with a grain of salt) provide their own supervisor for this reason among others (maybe the user wants a supervisor hierarchy for different parts of their application, etc.).  I would have something like `def KafkaEx.supervisor_spec(...)` that would return a  `Supervisor.Spec` that makes it relatively easy to work with.

**Comment by bjhaid** [comment]
> It might be a good idea to expose the KafkaEx.Supervisor restart strategy parameters to the application config.

I recommended same in this issue https://github.com/kafkaex/kafka_ex/issues/65 though the OP in the issue decided a different direction for a fix of his problem. You can also bump the `sync_timeout` to match how long you think it takes before it times out.

You can also use a `monitor` instead of linking that way your are not completely taken down and you can restart the worker, or alternatively make the linked process supervised and let it do the starting/linking of kafka_ex workers as part of its init.

I would implement exposing the worker spec to config today.
